pkg/alertmanager/alertmanager.go

```diff
@@ -237,7 +238,7 @@ func New(cfg *Config, reg *prometheus.Registry) (*Alertmanager, error) {
 		return nil, errors.Wrap(err, "failed to start state persister service")
 	}
 
-	am.pipelineBuilder = notify.NewPipelineBuilder(am.registry)
+	am.pipelineBuilder = notify.NewPipelineBuilder(am.registry, featurecontrol.NoopFlags{})
```

Comment:
New feature flag abstraction. In mimir for now we want all defaults for now; behave as if all experimental flags are off. If we want to support such flags in the future, we can easily map the configuration here to mimir cfg options.
https://github.com/prometheus/alertmanager/blob/main/featurecontrol/featurecontrol.go

There is one caveat:

`featurecontrol` defines a flag called `classicMatchersParsing`. This flag's name suggests keeping the _old_ behavior, not enabling something experimental. @grobinson-grafana is this correct? What happens if both `classicMatchersParsing` and `utf8MatchersParsing` are false, as they are in `NoopFlags`? Does it even make sense to have neither classic nor utf8 matcher parsing?

IMO if this doesn't make sense, we should consider changing `featurecontrol` such that it's consistent in the normal use case; `NoopFlags` should mean "all defaults, no experimental features." We can pass the classic flag in here if it's needed in the short term, but it feels dangerous to rely on this as we don't want to accidentally disable a piece of core AM functionality.

Reply:
I think just the `EnableReceiverNamesInMetrics()` feature flag is used here. The parsing feature flags are not used in Alertmanager server until (https://github.com/prometheus/alertmanager/pull/3567) is finished, and even then won't be used in `notify` package.

To answer the question about how the feature flags work:

1. No flags set (i.e. NoopFlags) means Alertmanager/amtool will use both parsers in a fallback mechanism.
2. A single flag set means exclusive use of one parser (i.e. classic or utf-8)
3. Both flags set is an [error](https://github.com/prometheus/alertmanager/blob/main/featurecontrol/featurecontrol.go#L110)

It might be worthwhile waiting until https://github.com/prometheus/alertmanager/pull/3567 is done before updating Alertmanager in Mimir, unless there is something critical that we need to update?

Reply:
> I think just the EnableReceiverNamesInMetrics() feature flag is used here. 

Yeah - I noticed that the flags weren't listened to. The danger here, IMO, is in future upgrades. If we pass in NoopFlags now, and in the future NoopFlags means something other than the standard behavior, the compiler will not tell us. We end up opting Mimir into experimental behavior, prematurely.

I don't think we want to do both parsers, until after the parser work is done. Especially since the fallback parser uses the new parser first. Sounds like I should pass in `classicMatchersParsing` here to safeguard us? It feels dangerous to me that the default option is to opt into incomplete behavior - it seems to defeat the point of the feature flag.

> It might be worthwhile waiting

I'd like to get in the habit of upgrading alertmanager more frequently. It hasn't been upgraded in a few months at this point. The longer we wait, the higher the risk at upgrade/rollout time.

Reply:
I think I'd quite like us to remove the `EnableReceiverNamesInMetrics` feature flag instead. It's been around for a while so it might be time to either make it permanent or remove it. The benefit of doing that is that we no longer need to pass feature flags to `NewPipelineBuilder` at all and then this issue becomes moot.

> Sounds like I should pass in classicMatchersParsing here to safeguard us? It feels dangerous to me that the default option is to opt into incomplete behavior - it seems to defeat the point of the feature flag.

You can do, but it shouldn't/won't have any effect, ever. The `matchers/compat` package should never be used in the notify package as there is just no need for them.

Reply:
Using the same flagset everywhere still future proofs us in case code does get shared in a future change.

Ideally, we'd even wire them up to Mimir flags and not prescribe a set, users can choose the flags at runtime. This is also outside the scope of this PR, IMO. For now, I want to keep this PR focused on selecting a good default flagset, that can be made configurable in the future.

Reply:
> Ideally, we'd even wire them up to Mimir flags and not prescribe a set, users can choose the flags at runtime.

Josh and I will be doing this as part of adding UTF-8 support to Alertmanager in the next couple of weeks. I think we will have both process-wide flags and also per tenant flags.

> For now, I want to keep this PR focused on selecting a good default flagset, that can be made configurable in the future.

I think `NoopFlags` are fine. These are just the default values for the struct, and are the default values in Alertmanager. Josh and I aren't going upgrade Alertmanager in Mimir without adding all the controls above. To be on the safe side we can use `classic-matchers-parsing` in the default. But this should be declared in `main` not in `pkg/alertmanager/alertmanager.go`.

Reply:
Mimir's `main` is very generic, it contains very little service-specific defaults like this.

These are flags for the per-tenant alertmanager, which is exactly what `alertmanager.go` implements - I feel the current file is the best place for it.

If we want per-tenant flags then it definitely cannot go in main. In fact, the current spot makes per-tenant flags the absolute easiest. This is the code that consumes the tenant config struct. I don't think we should move it.

Reply:
Oh I think I misunderstood :( I take it all back. The feature flags for `classicMatchersParsing` and `utf8MatchersParsing` are global (process-wide) at present due to the `matchers/compat` package, so even if configured per tenant each tenant would overwrite all other tenants when the feature flag is used. We're fixing that this week though, hence it might just be worthwhile waiting one more week for Josh and I to make those changes.

Reply:
Oh I think I misunderstood :( I take it all back. The feature flags for classicMatchersParsing and utf8MatchersParsing are global (process-wide) at present due to the matchers/compat package, so even if configured per tenant each tenant would overwrite all other tenants when the feature flag is used. We're fixing that this week though, hence it might just be worthwhile waiting one more week for Josh and I to make those changes.

If this is time urgent though I would just set the `classicMatchersParsing` as the default, as it's what Mimir uses in all previous versions. You might need to use a string literal as the original feature flag code in Alertmanager was not needed outside it's own CLI, hence it's all package private.

Reply:
Relevant PR: https://github.com/prometheus/alertmanager/pull/3596

Reply:
I just discussed with @stevesg an issue R.E https://github.com/prometheus/alertmanager/pull/3596 when it comes to parsing the Alertmanager configuration file. We cannot pass a per-tenant feature flag to `config.Load` as it cannot be propagated to `UnmarshalJSON/UnmarshalYAML` for `config.Matchers`. We will need to solve this somehow if we want per-tenant feature flags.

Reply:
That's fine - i still would expect that such a flag must eventually make its way to the per-tenant alertmanager. This does not seem to affect this PR.